File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.py
import sys
input = sys.stdin.readline

n,m = map(int,input().split())
board = [['0'] * (m+1)] + [['0'] + list(input().rstrip()) for _ in range(n)]
dp = [[0 for _ in range(m+1)] for _ in range(n+1)]
    
ans = 0
for i in range(1,n+1):
    for j in range(1,m+1):
        if board[i][j] == '1':
            dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1
            ans = max(ans, dp[i][j])

print(ans**2)
    

File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int dp[][] = new int[n][m];
		int ans = 0;
		for (int i = 0; i < n; i++) {
			String row = br.readLine();
			for (int j = 0; j < m; j++) {
				dp[i][j] = (int) row.charAt(j) - '0';
				if (dp[i][j] == 1 && j > 0 && i > 0) {
					dp[i][j] = Math.min(dp[i - 1][j - 1], Math.min(dp[i - 1][j], dp[i][j - 1])) + dp[i][j];
				}
				ans = Math.max(ans, dp[i][j]);
			}
		}
		System.out.println((int) Math.pow(ans, 2));
	}
}